src/odemis/acq/align/fastem.py

```diff
@@ -18,22 +20,282 @@
+
+                # TODO Connect the progress of the sub-future to the main future
+                # sf.add_update_callback(_pass_future_progress)
+                # sf.result()
+
+                # remove from list of calibrations when finished
+                self._calibrations_remaining.discard(calib)
+
+                # In case the calibrations was cancelled by a client, before the future returned,
+                # raise cancellation error.
+                if self._cancelled:
+                    raise CancelledError()
+
+                # Update the time left for the calibrations remaining
+                self._future.set_progress(end=time.time() + self.estimate_calibration_time())
+
+        except CancelledError:  # raised in try statement
+            logging.debug("Calibration was cancelled.")
+            raise
+        except Exception as ex:
+            raise ex
+        finally:
+            # Remove references to the calibrations once all calibrations are finished/cancelled.
+            self._calibrations_remaining.clear()
+            self._scanner.blanker.value = True  # always blank the beam to reduce beam damage on sample
+            logging.debug("Finish calibrations.")
+
+    def run_calibrations(self, calibration):
+        """
+        Run a calibration.
+        # Note: All calibrations can be run on bare scintillator.
+        """
+        if calibration == OPTICAL_AUTOFOCUS:
+            autofocus_multiprobe.run_autofocus(self._scanner, self._multibeam, self._descanner, self._detector,
+                                               self._dataflow, self._ccd, self._stage)
+
+        if calibration == SCAN_ROTATION_PREALIGN:
+            scan_rotation_pre_align.run_scan_rotation_pre_align(self._scanner, self._multibeam, self._descanner,
+                                                                self._detector, self._dataflow, self._ccd)
+
+        if calibration == SCAN_AMPLITUDE_PREALIGN:
+            scan_amplitude_pre_align.run_scan_amplitude_pre_align(self._scanner, self._multibeam, self._descanner,
+                                                                  self._detector, self._dataflow, self._ccd)
+
+        if calibration == DESCAN_GAIN_STATIC:
+            descan_gain.run_descan_gain_static(self._scanner, self._multibeam, self._descanner,
+                                               self._detector, self._dataflow, self._ccd)
+
+        if calibration == IMAGE_ROTATION_PREALIGN:
+            image_rotation_pre_align.run_image_rotation_pre_align(self._scanner, self._multibeam, self._descanner,
+                                                                  self._detector, self._dataflow, self._ccd,
+                                                                  self._det_rotator)
+
+        if calibration == IMAGE_TRANSLATION_PREALIGN:
+            image_translation_pre_align.run_image_translation_pre_align(self._scanner, self._multibeam,
```

Comment:
this calibration (and I think a few others) returns the calibrated value, which then still needs to be set on the hardware. How do you deal with that?

Reply:
Didn't think about this yet. Will do. I suggest similar as in the `run_calibrations.py` module?

Reply:
sounds good

Reply:
Let's discuss the details for this when I am back. It really depends also a bit if we want to allow to run subsets of the calibrations. Maybe that part is also easier when you had looked into the additional MD for the calibrated values next week.